### PR TITLE
Improve performance of ChatChannel and ChatMessage equality

### PR DESCRIPTION
**Pull request by laevandus**
### 🔗 Issue Links

Related to [PBE-5004](https://stream-io.atlassian.net/browse/PBE-5004)

### 🎯 Goal

* Reduce message and channel copies when calling `==`

### 📝 Summary

Time profiler shows in the SwiftUI SDK that when doing `==` checks a lot of time is spent in making copies of messages and channels. Changing the `==` to use guard let takes away a lot of this overhead.

### 🎨 Showcase

LazyCachedMapCollection's `==` is reduced from 118 ms > 11 ms  **>10** times improvement 

| Before | After |
| ------ | ----- |
|  Before   |  After  |

### 🧪 Manual Testing Notes

None

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] Changelog is updated with new localization keys
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

[PBE-5004]: https://stream-io.atlassian.net/browse/PBE-5004?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 # Upcoming
 
 ## StreamChat
+### ⚡ Performance
+- Improve performance of `ChatChannel` and `ChatMessage` quality checks [#3335](https://github.com/GetStream/stream-chat-swift/pull/3335)
```
Comment (martinmitrevski) [inline CHANGELOG.md:8]: you miss an "e" here (equality checks)

File: Sources/StreamChat/Models/Channel.swift [hunk 2]
```diff
@@ -237,27 +237,28 @@ extension ChatChannel: AnyChannel {}
...
-            lhs.truncatedAt == rhs.truncatedAt &&
-            lhs.watcherCount == rhs.watcherCount
+        guard lhs.cid == rhs.cid else { return false }
+        guard lhs.updatedAt == rhs.updatedAt else { return false }
+        guard lhs.lastMessageAt == rhs.lastMessageAt else { return false }
+        guard lhs.muteDetails == rhs.muteDetails else { return false }
+        guard lhs.reads == rhs.reads else { return false }
+        guard lhs.previewMessage == rhs.previewMessage else { return false }
+        guard lhs.name == rhs.name else { return false }
+        guard lhs.watcherCount == rhs.watcherCount else { return false }
```
Comment (martinmitrevski) [inline Sources/StreamChat/Models/Channel.swift:247]: I see you also changed the order, did that have an impact?

**Review by martinmitrevski — APPROVED** [review]
Great find ✅

File: Sources/StreamChat/Models/Channel.swift (hunk 2 above)
Comment (laevandus) [reply Sources/StreamChat/Models/Channel.swift:247]: Slightly re-ordered them depending on my gut feeling what could change more often. Nothing really noticeable. I have seen ownCapabilities being the most visible.

Comment (nuno-vieira) [inline Sources/StreamChat/Models/Channel.swift:247]: I'm not sure I fully understood the issue 🤔 Isn't the same code?

Comment (laevandus) [reply Sources/StreamChat/Models/Channel.swift:247]: Seems like compiler produces slightly different code where now it does not copy channels and messages so much as it used to be with a long expression.

**Pull request merged**